### DISCUSSION TITLE
IBLPrefilerContext.SpecularFilter VS Texture.generateMipmaps()

### DISCUSSION THREAD

**Discussion by ThomasGorisse** [google/filament#4665] [Q&A]
Hi,

I'd like to know what are the main differences between `IBLPrefilerContext.SpecularFilter.run()` and `Texture.generateMipmaps()`

I know that specular filter is working on the GPU which give me unusable results on a frame callback context. Even if the the base texture is 16x16x6, maybe because Filament doesn't have it's own dedicated coroutine context on my project for now but the Main one, every other rendering calls becomes laggy.

Texture.generateMimaps() is working smoothly but the result with a low resolution texture source becomes very rought when reflected on metal materials.

![Screenshot_20210923-080643.png](https://user-images.githubusercontent.com/6597529/134461280-fdadaa60-fcf1-4a6d-ba2c-1327db5f7568.png)

**Comment by romainguy** [comment]
> I'd like to know what are the main differences between IBLPrefilerContext.SpecularFilter.run() and Texture.generateMipmaps()

They do two different things. `generateMipmaps()` generates regular mipmaps, which are just downscaled versions of the original texture, using a box filter. `SpecularFilter` applies a filter based on the BRDF used for lighting.

> Even if the the base texture is 16x16x6, maybe because Filament doesn't have it's own dedicated coroutine context on my project 
> for now but the Main one, every other rendering calls becomes laggy.

Coroutines have nothing to do with this? Do you run `SpecularFilter` on every frame? If so, don't. It will take GPU time away from the rest of the app.

> Texture.generateMimaps() is working smoothly but the result with a low resolution texture source becomes very rought when 
> reflected on metal materials.

Since you start from a 16x16x16 texture, the results will look blurry no matter what. `SpecularFilter` will however be more correct than `generateMipmaps()` (note that `SpecularFilter` will use `generateMipmaps()` internally).

@pixelflinger is `IBLPrefilterContext` async or not? I don't remember.

**Comment by ThomasGorisse** [comment]
> Coroutines have nothing to do with this?

I was thinking that over app computations could interfere with the Filament rendering

> Do you run SpecularFilter on every frame? If so, don't. It will take GPU time away from the rest of the app.

I try to limit the `SpecularFilter` call to only light estimated changed frames but it's actually not so much relevant since even just a little camera move or estimation micro change needs a reflection update.
Do you know any manner to compare two texture images in order to decide whether on not to apply a new the `IndirectLight` reflections?

> Since you start from a 16x16x16 texture, the results will look blurry no matter what. SpecularFilter will however be more correct than generateMipmaps() (note that SpecularFilter will use generateMipmaps() internally).

The difference is not so big but in case of metal materials, specular filter gives more realistic result.
You can see the video links for the performance part.

### generatePrefilterMipmap

![Screenshot_20210923-134640](https://user-images.githubusercontent.com/6597529/134502756-60a5ffc0-6e4f-4832-a10b-bbf28444f90d.jpg)

Video : https://photos.app.goo.gl/7X57421FhqVEGhFc8

### SpecularFilter

![Screenshot_20210923-135023](https://user-images.githubusercontent.com/6597529/134502706-05a7927e-427d-44f6-9172-3d13d0bfa310.jpg)

Video: https://photos.app.goo.gl/Zse22xYWGF9TvexQ9
